src/core/lib/security/credentials/oauth2/oauth2_credentials.cc

```diff
@@ -472,6 +474,140 @@ grpc_call_credentials* grpc_google_refresh_token_credentials_create(
       .release();
 }
 
+//
+// STS credentials.
+//
+
+namespace {
+
+char* maybe_add_to_query(char* body, const char* field_name, char* field) {
+  if (field == nullptr || strlen(field) == 0) return body;
+  char* new_body;
+  gpr_asprintf(&new_body, "%s&%s=%s", body, field_name, field);
+  gpr_free(body);
+  return new_body;
+}
+
+class grpc_sts_token_fetcher_credentials
+    : public grpc_oauth2_token_fetcher_credentials {
+ public:
+  grpc_sts_token_fetcher_credentials(
+      grpc_uri* sts_url,  // Ownership transfered.
+      const grpc_sts_credentials_options* options)
+      : sts_url_(sts_url),
+        resource_(gpr_strdup(options->resource)),
+        audience_(gpr_strdup(options->audience)),
+        scope_(gpr_strdup(options->scope)),
+        requested_token_type_(gpr_strdup(options->requested_token_type)),
+        subject_token_(gpr_strdup(options->subject_token)),
+        subject_token_type_(gpr_strdup(options->subject_token_type)),
+        actor_token_(gpr_strdup(options->actor_token)),
+        actor_token_type_(gpr_strdup(options->actor_token_type)) {}
+
+  ~grpc_sts_token_fetcher_credentials() override { grpc_uri_destroy(sts_url_); }
+
+ protected:
+  void fetch_oauth2(grpc_credentials_metadata_request* metadata_req,
+                    grpc_httpcli_context* http_context,
+                    grpc_polling_entity* pollent,
+                    grpc_iomgr_cb_func response_cb,
+                    grpc_millis deadline) override {
+    grpc_http_header header = {(char*)"Content-Type",
+                               (char*)"application/x-www-form-urlencoded"};
+    grpc_httpcli_request request;
+    char* body = nullptr;
+    gpr_asprintf(&body, GRPC_STS_POST_MINIMAL_BODY_FORMAT_STRING,
+                 subject_token_.get(), subject_token_type_.get());
+    body = maybe_add_to_query(body, "resource", resource_.get());
```

Comment:
maybe_add_to_query seems not efficient. We have to create a new string and destroy the old one.
Consider use gpr_strvec_add and gpr_strvec_flatten. Mark Roth may give better recommendation.

Reply:
agree it's not efficient. Thanks for the pointer on strvec. Let me try it out.

Reply:
Looks like using strvec will give us the same number of malloc/free (actually more since we have to malloc/free the vector itself).  The mallocs will be smaller though and there will be less copies which I think is a win overall. Let me change the code.